src/command.rs

```diff
@@ -1,36 +1,327 @@
+    }
+
+    /// Whether or not the resulting executable should be sent to the 3DS with
+    /// `3dslink`.
+    pub fn should_link_to_device(&self) -> bool {
+        match self {
+            CargoCmd::Test(test) => !test.no_run,
+            CargoCmd::Run(_) => true,
+            _ => false,
+        }
+    }
+
+    pub const DEFAULT_MESSAGE_FORMAT: &str = "json-render-diagnostics";
+
+    pub fn extract_message_format(&mut self) -> Result<Option<String>, String> {
+        Self::extract_message_format_from_args(match self {
+            CargoCmd::Build(args) => &mut args.args,
+            CargoCmd::Run(run) => &mut run.cargo_args.args,
+            CargoCmd::Test(test) => &mut test.run_args.cargo_args.args,
+            CargoCmd::Passthrough(args) => args,
+        })
+    }
+
+    fn extract_message_format_from_args(
+        cargo_args: &mut Vec<String>,
+    ) -> Result<Option<String>, String> {
+        // Checks for a position within the args where '--message-format' is located
+        if let Some(pos) = cargo_args
+            .iter()
+            .position(|s| s.starts_with("--message-format"))
+        {
+            // Remove the arg from list so we don't pass anything twice by accident
+            let arg = cargo_args.remove(pos);
+
+            // Allows for usage of '--message-format=<format>' and also using space separation.
+            // Check for a '=' delimiter and use the second half of the split as the format,
+            // otherwise remove next arg which is now at the same position as the original flag.
+            let format = if let Some((_, format)) = arg.split_once('=') {
+                format.to_string()
+            } else {
+                // Also need to remove the argument to the --message-format option
+                cargo_args.remove(pos)
+            };
+
+            // Non-json formats are not supported so the executable exits.
+            if format.starts_with("json") {
+                Ok(Some(format))
+            } else {
+                Err(String::from(
+                    "error: non-JSON `message-format` is not supported",
+                ))
+            }
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl RemainingArgs {
+    /// Get the args to be passed to the executable itself (not `cargo`).
+    pub fn cargo_args(&self) -> &[String] {
+        self.split_args().0
+    }
+
+    /// Get the args to be passed to the executable itself (not `cargo`).
+    pub fn exe_args(&self) -> &[String] {
+        self.split_args().1
+    }
+
+    fn split_args(&self) -> (&[String], &[String]) {
+        if let Some(split) = self.args.iter().position(|s| s == "--") {
+            self.args.split_at(split + 1)
+        } else {
+            (&self.args[..], &[])
+        }
+    }
+}
+
+impl Run {
+    /// Get the args to pass to `3dslink` based on these options.
+    pub fn get_3dslink_args(&self) -> Vec<String> {
```

Comment:
Do you want to add tests for this? Not needed for the merge at least.

Reply:
Sure, I can add it whenever I try to add CI / other test assertions in a future PR